File: app/src/main/java/org/wikipedia/model/EnumCodeMap.java
package org.wikipedia.model;

import android.support.annotation.NonNull;
import android.util.SparseArray;

public class EnumCodeMap<T extends Enum<T> & EnumCode> {
    @NonNull private final SparseArray<T> map;

    public EnumCodeMap(@NonNull Class<T> enumeration) {
        map = codeToEnumMap(enumeration);
    }

    @NonNull public T get(int code) {
        T status = map.get(code);
        if (status == null) {
            throw new IllegalArgumentException("code=" + code);
        }
        return status;
    }

    @NonNull private SparseArray<T> codeToEnumMap(@NonNull Class<T> enumeration) {
        SparseArray<T> ret = new SparseArray<>();
        for (T value : enumeration.getEnumConstants()) {
            ret.put(value.code(), value);
        }
        return ret;
    }

    public int size() {
        return map.size();
    }
}

File: app/src/main/java/org/wikipedia/model/EnumCodeMap.kt
package org.wikipedia.model

import android.util.SparseArray
import androidx.core.util.valueIterator

class EnumCodeMap<T>(enumeration: Class<T>) where T : Enum<T>, T : EnumCode {
    private val map: SparseArray<T>

    init {
        map = codeToEnumMap(enumeration)
    }

    operator fun get(code: Int): T {
        return map.get(code) ?: throw IllegalArgumentException("code=$code")
    }

    private fun codeToEnumMap(enumeration: Class<T>): SparseArray<T> {
        val ret = SparseArray<T>()
        for (value in enumeration.enumConstants!!) {
            ret.put(value.code(), value)
        }
        return ret
    }

    fun size(): Int {
        return map.size()
    }

    fun valueIterator(): Iterator<T> {
        return map.valueIterator()
    }
}
